broadinstitute/single_cell_portal_core
# Pathway diagrams via related genes ideogram (SCP-5751)

This lets users see how their searched gene interacts in cell circuits, and overlays expression data on pathway diagrams.

https://github.com/user-attachments/assets/2e905a25-512d-4e48-873f-c33afcb9f4cc

### Impact
Previously, users got a brief text summary of how their searched gene interacts with adjacent genes in a biochemical pathway.  For example, "NR3C1 acts on CSN2" in a pathway like "Pregnancy and lactation".  This was useful -- some 5-8K single-gene searches per month (> 10%) derive from related genes ideogram -- but limited.  Genes work in directed graphs; the text only summarizes one adjacent gene among many connected genes that ultimately cause a biological condition.

Now, the pathway diagram is shown right in the embedding page, and it's interactive.  The searched gene and interacting gene are highlighted in the pathway while expression data loads.  The overlaid expression data comes from a dot plot.  Expression for any one annotation group can be depicted.  Like dot plots, genes with high mean scaled expression are colored red, medium purple, and low blue.  Percent of cells expression is conveyed via opacity: lower percent, lower opacity.

This gives deep, specific functional insight for transcriptomic data.

### Technical notes
The pathway diagrams are enabled by an upstream update in Ideogram.js.  Those retrieve diagram SVG from the WikiPathways GitHub repo by way of Pvjs, a pathway viewer JS component.  Pvjs also enables panning, zooming, and some diagram DOM API accessibility.

The expression overlay is enabled by SCP API endpoints for dotplots.  The default ideogram pathway diagram header is locally extended here to include an annotation label menu, and an info icon that briefly explains gene coloring and opacity in the pathway expression overlay.

A feature flag lets us disable the expression overlay in the pathway diagram, if we want.  Usage is already softly gated because pathway diagrams are only available when fairly deep into exploration: in a study, search 1 gene, hover over a purple gene marker, click a pathway.  This lets especially engaged users discover the new pathway expression functionality.  So keeping the functionality available by default seems to make sense.

### Known issues, limitations, and next steps
1. Around 2% of pathways do not render diagrams when clicked in the ideogram gene popover.
2. The framing UX seems viable, but a fuller integration into the many-gene display (e.g. 3+, with dot plots as default enabled tab) would be better.
3. The implementation is plain JS and sometimes inline CSS, rather than more SCP-conventional React and external CSS.  Much of the plain JS is a result of interacting with an external visualization library.  But refactoring some state management to React when upgrading the framing UX  (per 2) would be opportune.  
4. Diagram nodes indicate they're clickable due to Pvjs, but clicking them does nothing.  Ideally the cursor would be changed, or clicking would e.g. show expression metrics, gene information, etc.

### Test
To manually test:
1.  `yarn install`
2. `rails db:migrate`
3. Open local copy of "Cellular and transcriptional diversity over the course of human lactation" study
4. Search CSN2
6. Mouseover NR3C1 in related genes ideogram
7. In ideogram gene tooltip / popover, click "Mammary gland development: pregnancy and lactation - stage 3 of 4"
8. Confirm display of pathway diagram
9. Confirm expression overlay for pathway gene nodes renders after a brief wait
10. In "Expression in" menu, select another annotation label
11. Confirm expression overlay updates instantly

### Further details
More context is available in [2024-08-13 SCP demo](https://drive.google.com/file/d/12t_AgcHRknZepvsRw2fDrF8PS9vWq0cP/view?t=35s).  This satisfies SCP-5751, a Developer's Choice Days project.
